certcache.go

```diff
@@ -56,3 +108,245 @@ func (this *CertCache) ServeHTTP(resp http.ResponseWriter, req *http.Request, pa
 		http.NotFound(resp, req)
 	}
 }
+
+// If we've been unable to fetch a fresh OCSP response before expiry of the old
+// one, or, at server start-up, if we're unable to fetch a valid OCSP request at
+// all (either from disk or network), then return false. This signals to the
+// packager that it should not try to package anything; just proxy the content
+// unsigned. (sleevi #8)
+func (this *CertCache) IsHealthy() bool {
+	this.ocspMu.RLock()
+	defer this.ocspMu.RUnlock()
+	return this.isHealthy(this.ocsp)
+}
+
+func (this *CertCache) isHealthy(ocspResp []byte) bool {
+	if ocspResp == nil {
+		log.Println("OCSP response not yet fetched.")
+		return false
+	}
+	issuer := this.findIssuer()
+	if issuer == nil {
+		log.Println("Cannot find issuer certificate in CertFile.")
+		return false
+	}
+	resp, err := ocsp.ParseResponseForCert(ocspResp, this.certs[0], issuer)
+	if err != nil {
+		log.Println("Error parsing OCSP response:", err)
+		return false
+	}
+	if resp.NextUpdate.Before(time.Now()) {
+		log.Println("Cached OCSP is stale, NextUpdate:", resp.NextUpdate)
+		return false
+	}
+	return true
+}
+
+func (this *CertCache) maybeUpdateOCSP() error {
+	this.ocspMu.RLock()
+	var ocspUpdateAfter time.Time
+	ocsp, err := this.ocspFile.Read(context.Background(), this.shouldUpdateOCSP, func(orig []byte) []byte {
+		return this.updateOCSP(orig, &ocspUpdateAfter)
+	})
+	if err != nil {
+		return errors.Wrap(err, "Updating OCSP cache")
+	}
+	if len(ocsp) == 0 {
+		return errors.New("Missing OCSP response.")
+	}
+	if !this.isHealthy(ocsp) {
+		return errors.New("OCSP failed health check.")
+	}
+	// Convert read lock into write lock before updating ocsp and ocspUpdateAfter.
+	this.ocspMuMu.Lock()
+	this.ocspMu.RUnlock()
+	this.ocspMu.Lock()
+	this.ocspMuMu.Unlock()
+	this.ocsp = ocsp
+	if !ocspUpdateAfter.Equal(time.Time{}) {
+		// updateOCSP was called, and therefore a new HTTP cache expiry was set.
+		// TODO(twifkak): Write this to disk, so any replica can pick it up.
+		this.ocspUpdateAfter = ocspUpdateAfter
+	}
+	this.ocspMu.Unlock()
+	return nil
+}
+
+// Checks for OCSP updates every hour. Never terminates.
+func (this *CertCache) maintainOCSP(stop chan struct{}) {
+	// Only make one request per ocspCheckInterval, to minimize the impact
+	// on OCSP servers that are buckling under load (sleevi #5).
+	ticker := time.NewTicker(ocspCheckInterval)
```

Comment:
Could leverage the cron that rtv.go uses. Refactor to pull out the cron var so this new job can be added to it...

Reply:
I think this has a few benefits over the `robfig/cron` library:
 - One fewer dependency.
 - The syntax is checked at compile-time, rather than "@every 1h" which happens at run-time. That said: 1) The actual logic isn't checked, 2) If we need more complicated scheduling (e.g. "every tuesday") then cron is probably better, and 3) It looks like cron has a `ConstantDelaySchedule` type that address my complaint. So I'm on the fence on this one.
 - With this format, main.go and unit tests can start/stop only the schedules they care about. With cron, you can basically only add a bunch of funcs during initialization. If you want to change the funcs that run, you need to replace the whole Cron to clear it (AFAICT). But maybe that's good enough for tests.

WDYT?

Reply:
As long as we're consistent and use one way. So either hange this to cron or change rtv to use NewTicker (i.e. add TODO for whatever is decided)

I think we can get around the cron initialization/testing through clever coding.

Reply:
OK, I added a TODO to change rtv, as long as that's OK by you. Partly motivated by trying to do things "the Go way", and partly by not knowing what robfig/cron does (e.g. I started perusing and noticed that it eats panics from the jobs it runs).